### PR TITLE
Fix 208: support advanced TLC configuration files

### LINKED ISSUE #208: [BUG] Could not parse TLC configuration in BlockingQueue.cfg: string matching regex '\z' expected but '{' found
```bash
kuppe@mkuppe-z440:~/src/TLA/blockingqueue$ ../apalache/bin/apalache-mc check BlockingQueue.tla
# Tool home: /home/mkuppe/src/TLA/apalache
# Package:   /home/mkuppe/src/TLA/apalache/mod-distribution/target/apalache-pkg-0.7.3-SNAPSHOT-full.jar
# JVM args: -Xmx4096m -Djava.library.path=/home/mkuppe/src/TLA/apalache/3rdparty/lib  -DTLA-Library=/home/mkuppe/src/TLA/apalache/src/tla:/home/mkuppe/src/TLA/blockingqueue/
#
# APALACHE version 0.7.3-SNAPSHOT build v0.7.0-28-gdd171c8
#
# WARNING: This tool is in the experimental stage.
#          Please report bugs at: [https://github.com/informalsystems/apalache/issues]

Checker options: filename=BlockingQueue.tla, init=, next=, inv=   I@19:04:13.850
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.google.inject.internal.cglib.core.$ReflectUtils$1 (file:/home/mkuppe/src/TLA/apalache/mod-distribution/target/apalache-pkg-0.7.3-SNAPSHOT-full.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int,java.security.ProtectionDomain)
WARNING: Please consider reporting this to the maintainers of com.google.inject.internal.cglib.core.$ReflectUtils$1
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
PASS #0: SanyParser                                               I@19:04:14.139
Parsing file /home/mkuppe/src/TLA/blockingqueue/BlockingQueue.tla
Parsing file /tmp/Naturals.tla
Parsing file /tmp/Sequences.tla
Parsing file /tmp/FiniteSets.tla
Parsing file /home/mkuppe/src/TLA/blockingqueue/TLAPS.tla
PASS #1: ConfigurationPass                                        I@19:04:14.482
  > Loading TLC configuration from BlockingQueue.cfg              I@19:04:14.483
Configuration error (see the manual):   > Could not parse TLC configuration in BlockingQueue.cfg: string matching regex '\z' expected but '{' found E@19:04:14.523
It took me 0 days  0 hours  0 min  0 sec                          I@19:04:14.524
Total time: 0.727 sec                                             I@19:04:14.524
EXITCODE: ERROR (99)

mkuppe@mkuppe-z440:~/src/TLA/blockingqueue$ cat BlockingQueue.cfg 
\* SPECIFICATION
CONSTANTS
    BufCapacity = 3
    Producers = {p1,p2,p3,p4}
    Consumers = {c1,c2,c3}

SPECIFICATION FairSpec

INVARIANT Invariant
INVARIANT TypeInv

PROPERTY Starvation

mkuppe@mkuppe-z440:~/src/TLA/blockingqueue$ uname -a
Linux mkuppe-z440 5.4.0-42-generic #46-Ubuntu SMP Fri Jul 10 00:24:02 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

mkuppe@mkuppe-z440:~/src/TLA/blockingqueue$ java -version
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment (build 11.0.8+10-post-Ubuntu-0ubuntu120.04)
OpenJDK 64-Bit Server VM (build 11.0.8+10-post-Ubuntu-0ubuntu120.04, mixed mode, sharing)
```

```tla
--------------------------- MODULE BlockingQueue ---------------------------
EXTENDS Naturals, Sequences, FiniteSets

CONSTANTS Producers,   (* the (nonempty) set of producers                       *)
          Consumers,   (* the (nonempty) set of consumers                       *)
          BufCapacity  (* the maximum number of messages in the bounded buffer  *)

ASSUME Assumption ==
       /\ Producers # {}                      (* at least one producer *)
       /\ Consumers # {}                      (* at least one consumer *)
       /\ Producers \intersect Consumers = {} (* no thread is both consumer and producer *)
       /\ BufCapacity \in (Nat \ {0})         (* buffer capacity is at least 1 *)
       
-----------------------------------------------------------------------------

VARIABLES buffer, waitSet
vars == <<buffer, waitSet>>

RunningThreads == (Producers \cup Consumers) \ waitSet

NotifyOther(t) == 
          LET S == IF t \in Producers THEN waitSet \ Producers ELSE waitSet \ Consumers
          IN IF S # {}
             THEN \E x \in S : waitSet' = waitSet \ {x}
             ELSE UNCHANGED waitSet

(* @see java.lang.Object#wait *)
Wait(t) == /\ waitSet' = waitSet \cup {t}
           /\ UNCHANGED <<buffer>>
           
-----------------------------------------------------------------------------

Put(t, d) ==
/\ t \notin waitSet
/\ \/ /\ Len(buffer) < BufCapacity
      /\ buffer' = Append(buffer, d)
      /\ NotifyOther(t)
   \/ /\ Len(buffer) = BufCapacity
      /\ Wait(t)
      
Get(t) ==
/\ t \notin waitSet
/\ \/ /\ buffer # <<>>
      /\ buffer' = Tail(buffer)
      /\ NotifyOther(t)
   \/ /\ buffer = <<>>
      /\ Wait(t)

-----------------------------------------------------------------------------

(* Initially, the buffer is empty and no thread is waiting. *)
Init == /\ buffer = <<>>
        /\ waitSet = {}

(* Then, pick a thread out of all running threads and have it do its thing. *)
Next == \/ \E p \in Producers: Put(p, p) \* Add some data to buffer
        \/ \E c \in Consumers: Get(c)

-----------------------------------------------------------------------------

(* TLA+ is untyped, thus lets verify the range of some values in each state. *)
TypeInv == /\ buffer \in Seq(Producers)
           /\ Len(buffer) \in 0..BufCapacity
           /\ waitSet \in SUBSET (Producers \cup Consumers)

(* No Deadlock *)
Invariant == waitSet # (Producers \cup Consumers)

-----------------------------------------------------------------------------

MySeq(P) == UNION {[1..n -> P] : n \in 0..BufCapacity}

INSTANCE TLAPS

Spec == Init /\ [][Next]_vars

(* Whitelist all the known facts and definitions (except IInv below) *)
USE Assumption DEF vars, RunningThreads,
                   Init, Next, Spec,
                   Put, Get,
                   Wait, NotifyOther,
                   TypeInv, Invariant

\* TypeInv will be a conjunct of the inductive invariant, so prove it inductive.
\* An invariant I is inductive, iff Init => I and I /\ [Next]_vars => I. Note
\* though, that TypeInv itself won't imply Invariant though!  TypeInv alone
\* does not help us prove Invariant.
\* Luckily, TLAPS does not require us to decompose the proof into substeps. 
LEMMA TypeCorrect == Spec => []TypeInv
<1>1. Init => TypeInv BY SMT 
<1>2. TypeInv /\ [Next]_vars => TypeInv' BY SMT 
<1>. QED BY <1>1, <1>2, PTL

\* The naive thing to do is to check if the conjunct of TypeInv /\ Invariant
\* is inductive.
IInv == /\ TypeInv!2
        /\ TypeInv!3
        /\ Invariant
        \* When the buffer is empty, a consumer will be added to the waitSet.
        \* However, this does not crate a deadlock, because at least one producer
        \* will not be in the waitSet.
        /\ buffer = <<>> => \E p \in Producers : p \notin waitSet
        \* Vice versa, when buffer is full, a producer will be added to waitSet,
        \* but at least one consumer won't be in waitSet.
        /\ Len(buffer) = BufCapacity => \E c \in Consumers : c \notin waitSet

THEOREM DeadlockFreedom == Spec => []Invariant
<1>1. Init => IInv BY SMT DEF IInv
<1>2. IInv /\ [Next]_vars => IInv' BY DEF IInv
<1>3. IInv => Invariant  BY DEF IInv
<1>4. QED BY <1>1,<1>2,<1>3,PTL

MCIInv == TypeInv!1 /\ IInv

-----------------------------------------------------------------------------

PutEnabled == \A p \in Producers : ENABLED Put(p, p)

FairSpec == Spec /\ WF_vars(Next) 
                 /\ \A p \in Producers : WF_vars(Put(p, p)) 

(* All producers will continuously be serviced. For this to be violated,    *)
(* ASSUME Cardinality(Producers) > 1 has to hold (a single producer cannot  *)
(* starve itself).                                                          *)
Starvation == \A p \in Producers: []<>(<<Put(p, p)>>_vars)

=============================================================================
```

### PR DESCRIPTION
**Pull request by konnov**
Fix #208. This PR is larger than I expected. It took me a day of work instead of a few hours.

 - [x] detailed manual on TLC configuration options in `docs/tlc-config.md`
 - [x] extended parser for `CONSTANTS`
 - [x] ignoring unsupported features without failing on them
 - [x] fix the overrides in `ConfigurationPassImpl`, which was accidentally disabled by previous refactorings
 - [x] added integration tests for assignments and replacements

### CODE REVIEW

File: tla-import/src/main/scala/at/forsyte/apalache/io/tlc/config/TlcConfig.scala [hunk 1]
```diff
@@ -20,13 +25,60 @@ case class InitNextSpec(init: String, next: String) extends BehaviorSpec
...
+  def toTlaEx: TlaEx
+}
+
+/**
+  * A TLC model value, that is, a unique identifier that is treated as an uninterpreted constant.
+  * @param name the name of a model value
+  */
+case class ConfigModelValue(name: String) extends ConfigConstExpr {
+  override def toTlaEx: TlaEx = ValEx(TlaStr(name))
+}
```
Comment (andrey-kuprianov) [inline tla-import/src/main/scala/at/forsyte/apalache/io/tlc/config/TlcConfig.scala:50]: I see a problem here: from the user perspective, model values and string constants are different entities; However, as implemented, a model value A, and a string value "A" result in the same LIR expression. What about encoding model values with some special symbol, e.g. $?

```suggestion
case class ConfigModelValue(name: String) extends ConfigConstExpr {
  override def toTlaEx: TlaEx = ValEx(TlaStr(name))
}
```
```suggestion
case class ConfigModelValue(name: String) extends ConfigConstExpr {
  override def toTlaEx: TlaEx = ValEx(TlaStr("$" + name))
}
```

File: docs/tlc-config.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+// Recent feature: https://lamport.azurewebsites.net/tla/current-tools.pdf
+// APALACHE IGNORES THIS CONFIGURATION OPTION.
+Postcondition ::=
+    "POSTCONDITION" ident
+
+// Very recent feature: https://github.com/tlaplus/tlaplus/issues/485
+// APALACHE IGNORES THIS CONFIGURATION OPTION.
+Alias ::=
+    "ALIAS" ident
```
Comment (andrey-kuprianov) [inline docs/tlc-config.md:148]: I doubt it's a good idea to just silently ignore so many configuration parameters. The user will expect certain behavior, and the tool will probably just fail in another place, e.g. due to missing invariant, not explaining to the user that actually his PROPERTIES definition was ignored.

What about adding to the parser smth. like two string variables, warnings, and errors? And for whatever not supported, add lines to one of those? Later, when the parsed configuration is imported, check on these, and if errors are non-empty -> abort, if warnings are non-empty -> print them out.

**Review by andrey-kuprianov — CHANGES_REQUESTED** [review]
Great work, @konnov!

Only a couple of minor comments from my side; see above.
What would be also nice to have is some more tests, in particular to have tests that include the ignored config features.

File: docs/tlc-config.md (hunk 2 above)
Comment (konnov) [reply docs/tlc-config.md:148]: Well, `ConfigurationPassImpl` warns the user about the ignored properties. However, I do not want to make the configuration model too complex, just to complain about unsupported options later. We can add warnings right in the parser. Apart from `PROPERTY`, many other ignored options are TLC optimizations, which we will most probably never support.

File: tla-import/src/main/scala/at/forsyte/apalache/io/tlc/config/TlcConfig.scala (hunk 1 above)
Comment (konnov) [reply tla-import/src/main/scala/at/forsyte/apalache/io/tlc/config/TlcConfig.scala:50]: Good point, I will add a prefix like `ModelValue_`. I see model values more like a hack, as they are semantically just uninterpreted constants, not so different from strings in TLA+.

Comment (andrey-kuprianov) [reply tla-import/src/main/scala/at/forsyte/apalache/io/tlc/config/TlcConfig.scala:50]: agree; it's a hack:) but still good to differentiate between the two types, such that the user doesn't run into some unexpected errors.

File: docs/tlc-config.md (hunk 2 above)
Comment (andrey-kuprianov) [reply docs/tlc-config.md:148]: Right, warnings right in the parser is better, I only was not sure whether this is possible from the architectural point of view.

Btw, there is a warning issued about PROPERTIES in the importer, but this is already outdated, the leftover from the previous ConfigParser implementation, is it right? then it should be probably deleted?

**Comment by shonfeder** [comment]
Hmm. The merge conflicts in the integration tests should be easy to resolve, but that's a warning sign IMO that we should probably get this merged in lest we hit more troublesome conflicts.

Let me know if I can help here.

File: tla-import/src/main/scala/at/forsyte/apalache/io/tlc/config/TlcConfig.scala (hunk 1 above)
Comment (andrey-kuprianov) [reply tla-import/src/main/scala/at/forsyte/apalache/io/tlc/config/TlcConfig.scala:50]: Returning back to this: I think adding the `ModelValue_` prefix would be an overkill. Imagine Apalache returning counterexamples like `your_set = {"ModelValue_A", "ModelValue_B", "ModelValue_C"}` 

I would really prefer to have strings and model values differentiated, but in a non-overwhelming way, like with a small prefix: `$`, `#`, `_` ... But even leaving the strings clashing with the model values would be still better than the `ModelValue_` prefix.

File: docs/tlc-config.md (hunk 2 above)
Comment (lemmy) [reply docs/tlc-config.md:148]: Don't forget about [typed model values](https://tla.msr-inria.inria.fr/tlatoolbox/doc/model/model-values.html).

Comment (konnov) [reply docs/tlc-config.md:148]: I have added warnings for the options that are completely ignored by the parser such as "VIEW", "SYMMETRY", etc. We might want to extend the parser later, to propagate these options, but I do not see any value in doing that for now. The "PROPERTIES" are propagated and then ignored by the model checker with a warning, as we like to implement liveness checking later.

Comment (konnov) [reply docs/tlc-config.md:148]: > Don't forget about typed model values.

From the perspective of the config parser, there is no difference between model values and typed model values, right?

File: tla-import/src/main/scala/at/forsyte/apalache/io/tlc/config/TlcConfig.scala (hunk 1 above)
Comment (konnov) [reply tla-import/src/main/scala/at/forsyte/apalache/io/tlc/config/TlcConfig.scala:50]: We are already using `$` when renaming variables. I believe lots of unusual ASCII symbols confuse users a lot. Here it is clear that the string is coming from the config as a model value :)

**Pull request merged**